docs/proposals/rescheduling-for-critical-pods.md

```diff
@@ -0,0 +1,116 @@
+
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Rescheduler: guaranteed scheduling of critical addons
+
+## Motivation
+
+In addition to Kubernetes core components like api-server, scheduler, controller-manager running on a master machine
+there is a bunch of addons which due to various reasons have to run on a regular cluster node, not the master.
+Some of them are critical to have fully functional cluster: Heapster, DNS, UI. Users can break their cluster
+by evicting a critical addon (either manually or as a side effect of an other operation like upgrade)
+which possibly can become pending (for example when the cluster is highly utilized).
+To avoid such situation we want to have a mechanism which guarantees that
+critical addons are scheduled assuming the cluster is big enough.
+This possibly may affect other pods (including production user’s applications).
+
+## Design
+
+Rescheduler will ensure that critical addons are always scheduled.
+In the first version it will implement only this policy, but later we may want to introduce other policies.
+It will be a standalone component running on master machine similarly to scheduler.
+Those components will share common logic (initially rescheduler will in fact import some of scheduler packages).
+
+### Guaranteed scheduling of critical addons
+
+Rescheduler will observe critical addons (running in `kube-system` namespace
+with annotation `alpha.kubernetes.io/service-type` set to `CRITICAL`).
+If one of them is marked by scheduler as unschedulable (pod condition `PodScheduled` set to `false`, the reason set to `Unschedulable`)
+the component will try to find a space for the addon by evicting some pods and then the scheduler will schedule the addon.
+
+#### Scoring nodes
+
+Initially we want to choose a random node with enough capacity
+(chosen as described in [Evicting pods](rescheduling-for-critical-pods.md#evicting-pods)) to schedule given addons.
+Later we may want to introduce some heuristic:
+* minimize number of evicted pods with violation of disruption budget or shortened termination grace period
+* minimize number of affected pods by choosing a node on which we have to evict less pods
```

Comment:
> Is there a hurry for getting this feature as proposed in to solve something specific or just to get feedback from the community and collect more requirements before we zero in on the design of the more generic priority based pre-emption ?

Yes, unfortunately there is a hurry -- some users have had DNS or Heapster or both (I can't remember) fail to schedule after being evicted (I forget the reason for the eviction) because their cluster was full of regular pods. As you can imagine, bad things happen when DNS fails to schedule. Heapster (and the UI) are marginally less critical, but it's stillpretty bad when you lose the ability to monitor the cluster.

My hope is that post-1.4 we can implement proper preemption in the scheduler (or some combination of the scheduler and kubelet, as @vishh proposed), and remove the logic described in this proposal while keeping the core intact as the framework for implementing the things we had been planning to do with rescheduler (e.g. move pods from heavily-utilized to under-utilized nodes). We should have a discussion on the right way to do proper priorities and preemptions before implementing it. (Some of the issues are described [here](https://github.com/kubernetes/kubernetes/blob/master/docs/proposals/rescheduling.md) but it's definitely not a full design and deserves a full design doc.)